README.md
# Emperror: Errors [![Mentioned in Awesome Go](https://awesome.re/mentioned-badge-flat.svg)](https://github.com/avelino/awesome-go#error-handling)

[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/emperror/errors/CI?style=flat-square)](https://github.com/emperror/errors/actions?query=workflow%3ACI)
[![Codecov](https://img.shields.io/codecov/c/github/emperror/errors?style=flat-square)](https://codecov.io/gh/emperror/errors)
[![Go Report Card](https://goreportcard.com/badge/emperror.dev/errors?style=flat-square)](https://goreportcard.com/report/emperror.dev/errors)
![Go Version](https://img.shields.io/badge/go%20version-%3E=1.12-61CFDD.svg?style=flat-square)
[![PkgGoDev](https://pkg.go.dev/badge/mod/emperror.dev/errors)](https://pkg.go.dev/mod/emperror.dev/errors)
[![FOSSA Status](https://app.fossa.com/api/projects/custom%2B8125%2Femperror.dev%2Ferrors.svg?type=shield)](https://app.fossa.com/projects/custom%2B8125%2Femperror.dev%2Ferrors?ref=badge_shield)


**Drop-in replacement for the standard library `errors` package and [github.com/pkg/errors](https://github.com/pkg/errors).**

This is a single, lightweight library merging the features of standard library `errors` package
and [github.com/pkg/errors](https://github.com/pkg/errors). It also backports a few features
(like Go 1.13 error handling related features).

Standard library features:
- `New` creates an error with stack trace
- `Unwrap` supports both Go 1.13 wrapper (`interface { Unwrap() error }`) and **pkg/errors** causer (`interface { Cause() error }`) interface
- Backported `Is` and `As` functions

[github.com/pkg/errors](https://github.com/pkg/errors) features:
- `New`, `Errorf`, `WithMessage`, `WithMessagef`, `WithStack`, `Wrap`, `Wrapf` functions behave the same way as in the original library
- `Cause` supports both Go 1.13 wrapper (`interface { Unwrap() error }`) and **pkg/errors** causer (`interface { Cause() error }`) interface

Additional features:
- `NewPlain` creates a new error without any attached context, like stack trace
- `Sentinel` is a shorthand type for creating [constant error](https://dave.cheney.net/2016/04/07/constant-errors)
- `WithStackDepth` allows attaching stack trace with a custom caller depth
- `WithStackDepthIf`, `WithStackIf`, `WrapIf`, `WrapIff` only annotate errors with a stack trace if there isn't one already in the error chain
- Multi error aggregating multiple errors into a single value
- `NewWithDetails`, `WithDetails` and `Wrap*WithDetails` functions to add key-value pairs to an error
- Match errors using the `match` package


## Installation

```bash
go get emperror.dev/errors
```


## Usage

```go
package main

import "emperror.dev/errors"

// ErrSomethingWentWrong is a sentinel error which can be useful within a single API layer.
const ErrSomethingWentWrong = errors.Sentinel("something went wrong")

// ErrMyError is an error that can be returned from a public API.
type ErrMyError struct {
	Msg string
}

func (e ErrMyError) Error() string {
	return e.Msg
}

func foo() error {
	// Attach stack trace to the sentinel error.
	return errors.WithStack(ErrSomethingWentWrong)
}

func bar() error {
	return errors.Wrap(ErrMyError{"something went wrong"}, "error")
}

func main() {
	if err := foo(); err != nil {
		if errors.Cause(err) == ErrSomethingWentWrong { // or errors.Is(ErrSomethingWentWrong)
			// handle error
		}
	}

	if err := bar(); err != nil {
		if errors.As(err, &ErrMyError{}) {
			// handle error
		}
	}
}
```

Match errors:

```go
package main

import (
    "emperror.dev/errors"
    "emperror.dev/errors/match"
)

// ErrSomethingWentWrong is a sentinel error which can be useful within a single API layer.
const ErrSomethingWentWrong = errors.Sentinel("something went wrong")

type clientError interface{
    ClientError() bool
}

func foo() error {
	// Attach stack trace to the sentinel error.
	return errors.WithStack(ErrSomethingWentWrong)
}

func main() {
    var ce clientError
    matcher := match.Any{match.As(&ce), match.Is(ErrSomethingWentWrong)}

	if err := foo(); err != nil {
		if matcher.MatchError(err) {
			// you can use matchers to write complex conditions for handling (or not) an error
            // used in emperror
		}
	}
}
```


## Development

When all coding and testing is done, please run th e test suite:

```bash
./pleasew lint
./pleasew gotest
```


## License

The MIT License (MIT). Please see [License File](LICENSE) for more information.

Certain parts of this library are inspired by (or entirely copied from) various third party libraries.
Their licenses can be found in the [Third Party License File](LICENSE_THIRD_PARTY).

[![FOSSA Status](https://app.fossa.com/api/projects/custom%2B8125%2Femperror.dev%2Ferrors.svg?type=large)](https://app.fossa.com/projects/custom%2B8125%2Femperror.dev%2Ferrors?ref=badge_large)


CHANGELOG.md
# Change Log


All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [Unreleased]

### Changed

- Update dependencies


## [0.7.0] - 2020-01-13

### Changed

- Updated dependencies ([github.com/pkg/errors](https://github.com/pkg/errors))
- Use `WithMessage` and `WithMessagef` from [github.com/pkg/errors](https://github.com/pkg/errors)


## [0.6.0] - 2020-01-09

### Changed

- Updated dependencies


## [0.5.2] - 2020-01-06

### Changed

- `match`: exported `ErrorMatcherFunc`


## [0.5.1] - 2020-01-06

### Fixed

- `match`: race condition in `As`


## [0.5.0] - 2020-01-06

### Added

- `match` package for matching errors


## [0.4.3] - 2019-09-05

### Added

- `Sentinel` error type for creating [constant error](https://dave.cheney.net/2016/04/07/constant-errors)


## [0.4.2] - 2019-07-19

### Added

- `NewWithDetails` function to create a new error with details attached


## [0.4.1] - 2019-07-17

### Added

- `utils/keyval` package to work with key-value pairs.


## [0.4.0] - 2019-07-17

### Added

- Error details


## [0.3.0] - 2019-07-14

### Added

- Multi error
- `UnwrapEach` function


## [0.2.0] - 2019-07-12

### Added

- `*If` functions that only annotate an error with a stack trace if there isn't one already in the error chain


## [0.1.0] - 2019-07-12

- Initial release


[Unreleased]: https://github.com/emperror/errors/compare/v0.7.0...HEAD
[0.7.0]: https://github.com/emperror/errors/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/emperror/errors/compare/v0.5.2...v0.6.0
[0.5.2]: https://github.com/emperror/errors/compare/v0.5.1...v0.5.2
[0.5.1]: https://github.com/emperror/errors/compare/v0.5.0...v0.5.1
[0.5.0]: https://github.com/emperror/errors/compare/v0.4.3...v0.5.0
[0.4.3]: https://github.com/emperror/errors/compare/v0.4.2...v0.4.3
[0.4.2]: https://github.com/emperror/errors/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/emperror/errors/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/emperror/errors/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/emperror/errors/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/emperror/errors/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/emperror/errors/compare/v0.0.0...v0.1.0


wrap_test.go
// Copyright 2019 The Go Authors. All rights reserved.
// Use of this source code (or at least parts of it) is governed by a BSD-style
// license that can be found in the LICENSE_THIRD_PARTY file.

package errors_test

import (
	"io"
	"reflect"
	"testing"

	"emperror.dev/errors"
)

func TestUnwrap(t *testing.T) {
	err1 := errors.NewPlain("1")
	erra := wrapped{"wrap 2", err1}

	testCases := []struct {
		err  error
		want error
	}{
		{nil, nil},
		{wrapped{"wrapped", nil}, nil},
		{caused{"wrapped", nil}, nil},
		{err1, nil},
		{erra, err1},
		{wrapped{"wrap 3", erra}, erra},
		{caused{"wrap 3", erra}, erra},
	}
	for _, tc := range testCases {
		if got := errors.Unwrap(tc.err); got != tc.want {
			t.Errorf("Unwrap(%v) = %v, want %v", tc.err, got, tc.want)
		}
	}
}

type wrapped struct {
	msg string
	err error
}

func (e wrapped) Error() string { return e.msg }

func (e wrapped) Unwrap() error { return e.err }

type caused struct {
	msg string
	err error
}

func (e caused) Error() string { return e.msg }

func (e caused) Cause() error { return e.err }

func TestUnwrapEach(t *testing.T) {
	err := errors.WithMessage(
		errors.WithMessage(
			errors.WithMessage(
				errors.New("level 0"),
				"level 1",
			),
			"level 2",
		),
		"level 3",
	)

	var i int
	fn := func(err error) bool {
		i++

		return true
	}

	errors.UnwrapEach(err, fn)

	if got, want := i, 5; got != want {
		t.Errorf("error chain length does not match the expected one\nactual:   %d\nexpected: %d", got, want)
	}
}

func TestUnwrapEach_BreakTheLoop(t *testing.T) {
	err := errors.WithMessage(
		errors.WithMessage(
			errors.WithMessage(
				errors.New("level 0"),
				"level 1",
			),
			"level 2",
		),
		"level 3",
	)

	var i int
	fn := func(err error) bool {
		i++

		return !(i > 2)
	}

	errors.UnwrapEach(err, fn)

	if got, want := i, 3; got != want {
		t.Errorf("error chain length does not match the expected one\nactual:   %d\nexpected: %d", got, want)
	}
}

func TestUnwrapEach_NilError(t *testing.T) {
	var i int
	fn := func(err error) bool {
		i++

		return !(i > 2)
	}

	errors.UnwrapEach(nil, fn)

	if got, want := i, 0; got != want {
		t.Errorf("error chain length does not match the expected one\nactual:   %d\nexpected: %d", got, want)
	}
}

type nilError struct{}

func (nilError) Error() string { return "nil error" }

func TestCause(t *testing.T) {
	x := errors.NewPlain("error")
	tests := []struct {
		err  error
		want error
	}{
		{
			// nil error is nil
			err:  nil,
			want: nil,
		},
		{
			// explicit nil error is nil
			err:  (error)(nil),
			want: nil,
		},
		{
			// typed nil is nil
			err:  (*nilError)(nil),
			want: (*nilError)(nil),
		},
		{
			// uncaused error is unaffected
			err:  io.EOF,
			want: io.EOF,
		},
		{
			// wrapped error returns cause
			err:  wrapped{"ignored", io.EOF},
			want: io.EOF,
		},
		{
			// caused error returns cause
			err:  caused{"ignored", io.EOF},
			want: io.EOF,
		},
		{
			err:  x, // return from errors.New
			want: x,
		},
		{
			errors.WithMessage(nil, "whoops"),
			nil,
		},
		{
			errors.WithMessage(io.EOF, "whoops"),
			io.EOF,
		},
		{
			errors.WithMessagef(nil, "%s", "whoops"),
			nil,
		},
		{
			errors.WithMessagef(io.EOF, "%s", "whoops"),
			io.EOF,
		},
		{
			errors.WithStack(nil),
			nil,
		},
		{
			errors.WithStack(io.EOF),
			io.EOF,
		},
		{
			errors.WithStackDepth(nil, 0),
			nil,
		},
		{
			errors.WithStackDepth(io.EOF, 0),
			io.EOF,
		},
	}

	for i, tt := range tests {
		got := errors.Cause(tt.err)
		if !reflect.DeepEqual(got, tt.want) {
			t.Errorf("test %d: got %#v, want %#v", i+1, got, tt.want)
		}
	}
}


stack_test.go
// Copyright 2015 Dave Cheney <dave@cheney.net>. All rights reserved.
// Use of this source code (or at least parts of it) is governed by a BSD-style
// license that can be found in the LICENSE_THIRD_PARTY file.

package errors

import (
	"runtime"
	"testing"
)

func TestStackTrace(t *testing.T) {
	tests := []struct {
		stack *stack
		want  []string
	}{
		{
			callers(0),
			[]string{
				"emperror.dev/errors.TestStackTrace\n" +
					"\t.+/stack_test.go:18",
			},
		},
	}
	for i, tt := range tests {
		st := tt.stack.StackTrace()

		for j, want := range tt.want {
			testFormatRegexp(t, i, st[j], "%+v", want)
		}
	}
}

func stackTrace() StackTrace {
	const depth = 8

	var pcs [depth]uintptr

	n := runtime.Callers(1, pcs[:])

	var st stack = pcs[0:n]

	return st.StackTrace()
}

func TestStackTraceFormat(t *testing.T) {
	tests := []struct {
		StackTrace
		format string
		want   string
	}{{
		nil,
		"%s",
		`\[\]`,
	}, {
		nil,
		"%v",
		`\[\]`,
	}, {
		nil,
		"%+v",
		"",
	}, {
		nil,
		"%#v",
		`\[\]errors.Frame\(nil\)`,
	}, {
		make(StackTrace, 0),
		"%s",
		`\[\]`,
	}, {
		make(StackTrace, 0),
		"%v",
		`\[\]`,
	}, {
		make(StackTrace, 0),
		"%+v",
		"",
	}, {
		make(StackTrace, 0),
		"%#v",
		`\[\]errors.Frame{}`,
	}, {
		stackTrace()[:2],
		"%s",
		`\[stack_test.go stack_test.go\]`,
	}, {
		stackTrace()[:2],
		"%v",
		`\[stack_test.go:39 stack_test.go:88\]`,
	}, {
		stackTrace()[:2],
		"%+v",
		"\n" +
			"emperror.dev/errors.stackTrace\n" +
			"\t.+/stack_test.go:39\n" +
			"emperror.dev/errors.TestStackTraceFormat\n" +
			"\t.+/stack_test.go:92",
	}, {
		stackTrace()[:2],
		"%#v",
		`\[\]errors.Frame{stack_test.go:39, stack_test.go:100}`,
	}}

	for i, tt := range tests {
		testFormatRegexp(t, i, tt.StackTrace, tt.format, tt.want)
	}
}


errors_test.go
package errors

import (
	"testing"
)

func checkErrorMessage(t *testing.T, err error, message string) {
	t.Helper()

	if got, want := err.Error(), message; got != want {
		t.Errorf("error message does not match the expectd\nactual:   %s\nexpected: %s", got, want)
	}
}

func checkFormat(t *testing.T, err error, formats map[string][]string) {
	t.Helper()

	i := 1

	for format, want := range formats {
		testFormatCompleteCompare(t, i, err, format, want, true)

		i++
	}
}

func checkErrorNil(t *testing.T, err error) {
	t.Helper()

	if err != nil {
		t.Errorf("nil error is expected to result in nil\nactual: %#v", err)
	}
}

func checkUnwrap(t *testing.T, err error, origErr error) {
	t.Helper()

	if err, ok := err.(interface{ Unwrap() error }); ok {
		if got, want := err.Unwrap(), origErr; got != want {
			t.Errorf("error does not match the expected one\nactual:   %#v\nexpected: %#v", got, want)
		}
	} else {
		t.Fatal("error does not implement the wrapper (interface{ Unwrap() error}) interface")
	}

	if err, ok := err.(interface{ Cause() error }); ok {
		if got, want := err.Cause(), origErr; got != want {
			t.Errorf("error does not match the expected one\nactual:   %#v\nexpected: %#v", got, want)
		}
	} else {
		t.Fatal("error does not implement the causer (interface{ Cause() error}) interface")
	}
}


wrap.go
package errors

// Unwrap returns the result of calling the Unwrap method on err, if err implements
// Unwrap. Otherwise, Unwrap returns nil.
//
// It supports both Go 1.13 Unwrap and github.com/pkg/errors.Causer interfaces
// (the former takes precedence).
func Unwrap(err error) error {
	switch e := err.(type) {
	case interface{ Unwrap() error }:
		return e.Unwrap()

	case interface{ Cause() error }:
		return e.Cause()
	}

	return nil
}

// Cause returns the last error (root cause) in an err's chain.
// If err has no chain, it is returned directly.
//
// It supports both Go 1.13 errors.Wrapper and github.com/pkg/errors.Causer interfaces
// (the former takes precedence).
func Cause(err error) error {
	for {
		cause := Unwrap(err)
		if cause == nil {
			break
		}

		err = cause
	}

	return err
}


stack.go
// Copyright 2015 Dave Cheney <dave@cheney.net>. All rights reserved.
// Use of this source code (or at least parts of it) is governed by a BSD-style
// license that can be found in the LICENSE_THIRD_PARTY file.

package errors

import (
	"fmt"
	"runtime"

	"github.com/pkg/errors"
)

// StackTrace is stack of Frames from innermost (newest) to outermost (oldest).
//
// It is an alias of the same type in github.com/pkg/errors.
type StackTrace = errors.StackTrace

// Frame represents a program counter inside a stack frame.
// For historical reasons if Frame is interpreted as a uintptr
// its value represents the program counter + 1.
//
// It is an alias of the same type in github.com/pkg/errors.
type Frame = errors.Frame

// stack represents a stack of program counters.
//
// It is a duplicate of the same (sadly unexported) type in github.com/pkg/errors.
type stack []uintptr

// nolint: gocritic
func (s *stack) Format(st fmt.State, verb rune) {
	switch verb {
	case 'v':
		switch {
		case st.Flag('+'):
			for _, pc := range *s {
				f := Frame(pc)
				fmt.Fprintf(st, "\n%+v", f)
			}
		}
	}
}

func (s *stack) StackTrace() StackTrace {
	f := make([]Frame, len(*s))
	for i := 0; i < len(f); i++ {
		f[i] = Frame((*s)[i])
	}
	return f
}

// callers is based on the function with the same name in github.com/pkg/errors,
// but accepts a custom depth (useful to customize the error constructor caller depth).
func callers(depth int) *stack {
	const maxDepth = 32
	var pcs [maxDepth]uintptr
	n := runtime.Callers(2+depth, pcs[:])
	var st stack = pcs[0:n]
	return &st
}


match/match.go
package match

import (
	"reflect"

	"emperror.dev/errors"
)

// ErrorMatcher checks if an error matches a certain condition.
type ErrorMatcher interface {
	// MatchError checks if err matches a certain condition.
	MatchError(err error) bool
}

// ErrorMatcherFunc turns a plain function into an ErrorMatcher if it's definition matches the interface.
type ErrorMatcherFunc func(err error) bool

// MatchError calls the underlying function to check if err matches a certain condition.
func (fn ErrorMatcherFunc) MatchError(err error) bool {
	return fn(err)
}

// Any matches an error if any of the underlying matchers match it.
type Any []ErrorMatcher

// MatchError calls underlying matchers with err.
// If any of them matches err it returns true, otherwise false.
func (m Any) MatchError(err error) bool {
	for _, matcher := range m {
		if matcher.MatchError(err) {
			return true
		}
	}

	return false
}

// All matches an error if all of the underlying matchers match it.
type All []ErrorMatcher

// MatchError calls underlying matchers with err.
// If all of them matches err it returns true, otherwise false.
func (m All) MatchError(err error) bool {
	for _, matcher := range m {
		if !matcher.MatchError(err) {
			return false
		}
	}

	return true
}

// Is returns an error matcher that determines matching by calling errors.Is.
func Is(target error) ErrorMatcher {
	return ErrorMatcherFunc(func(err error) bool {
		return errors.Is(err, target)
	})
}

// As returns an error matcher that determines matching by calling errors.As.
func As(target interface{}) ErrorMatcher {
	if target == nil {
		panic("errors: target cannot be nil")
	}

	val := reflect.ValueOf(target)
	typ := val.Type()

	if typ.Kind() != reflect.Ptr || val.IsNil() {
		panic("errors: target must be a non-nil pointer")
	}

	if e := typ.Elem(); e.Kind() != reflect.Interface && !e.Implements(errorType) {
		panic("errors: *target must be interface or implement error")
	}

	tar := reflect.New(typ).Interface()

	return ErrorMatcherFunc(func(err error) bool {
		target := tar

		return errors.As(err, &target)
	})
}

var errorType = reflect.TypeOf((*error)(nil)).Elem()


errors_multi.go
package errors

import (
	"go.uber.org/multierr"
)

// Combine combines the passed errors into a single error.
//
// If zero arguments were passed or if all items are nil, a nil error is
// returned.
//
// If only a single error was passed, it is returned as-is.
//
// Combine omits nil errors so this function may be used to combine
// together errors from operations that fail independently of each other.
//
// 	errors.Combine(
// 		reader.Close(),
// 		writer.Close(),
// 		pipe.Close(),
// 	)
//
// If any of the passed errors is already an aggregated error, it will be flattened along
// with the other errors.
//
// 	errors.Combine(errors.Combine(err1, err2), err3)
// 	// is the same as
// 	errors.Combine(err1, err2, err3)
//
// The returned error formats into a readable multi-line error message if
// formatted with %+v.
//
// 	fmt.Sprintf("%+v", errors.Combine(err1, err2))
func Combine(errors ...error) error {
	return multierr.Combine(errors...)
}

// Append appends the given errors together. Either value may be nil.
//
// This function is a specialization of Combine for the common case where
// there are only two errors.
//
// 	err = errors.Append(reader.Close(), writer.Close())
//
// The following pattern may also be used to record failure of deferred
// operations without losing information about the original error.
//
// 	func doSomething(..) (err error) {
// 		f := acquireResource()
// 		defer func() {
// 			err = errors.Append(err, f.Close())
// 		}()
func Append(left error, right error) error {
	return multierr.Append(left, right)
}

// GetErrors returns a slice containing zero or more errors that the supplied
// error is composed of. If the error is nil, the returned slice is empty.
//
// 	err := errors.Append(r.Close(), w.Close())
// 	errors := errors.GetErrors(err)
//
// If the error is not composed of other errors, the returned slice contains
// just the error that was passed in.
//
// Callers of this function are free to modify the returned slice.
func GetErrors(err error) []error {
	return multierr.Errors(err)
}


utils/keyval/to_map.go
package keyval

import (
	"fmt"
	"reflect"
)

// ToMap creates a map of key-value pairs from a variadic key-value pair slice.
//
// The implementation bellow is from go-kit's JSON logger.
func ToMap(kvs []interface{}) map[string]interface{} {
	m := map[string]interface{}{}

	if len(kvs) == 0 {
		return m
	}

	if len(kvs)%2 == 1 {
		kvs = append(kvs, nil)
	}

	for i := 0; i < len(kvs); i += 2 {
		merge(m, kvs[i], kvs[i+1])
	}

	return m
}

func merge(dst map[string]interface{}, k, v interface{}) {
	var key string

	switch x := k.(type) {
	case string:
		key = x
	case fmt.Stringer:
		key = safeString(x)
	default:
		key = fmt.Sprint(x)
	}

	dst[key] = v
}

func safeString(str fmt.Stringer) (s string) {
	defer func() {
		if panicVal := recover(); panicVal != nil {
			if v := reflect.ValueOf(str); v.Kind() == reflect.Ptr && v.IsNil() {
				s = "NULL"
			} else {
				panic(panicVal)
			}
		}
	}()

	s = str.String()

	return
}


errors.go
/*
	Package errors is a drop-in replacement for the standard errors package and github.com/pkg/errors.


	Overview

	This is a single, lightweight library merging the features of standard library `errors` package
	and https://github.com/pkg/errors. It also backports a few features
	(like Go 1.13 error handling related features).


	Printing errors

	If not stated otherwise, errors can be formatted with the following specifiers:
		%s	error message
		%q	double-quoted error message
		%v	error message in default format
		%+v	error message and stack trace
*/
package errors // import "emperror.dev/errors"

import (
	"fmt"
	"io"

	"github.com/pkg/errors"
)

// NewPlain returns a simple error without any annotated context, like stack trace.
// Useful for creating sentinel errors and in testing.
//
// 	var ErrSomething = errors.NewPlain("something went wrong")
func NewPlain(message string) error {
	return &plainError{message}
}

// plainError is a trivial implementation of error.
type plainError struct {
	msg string
}

func (e *plainError) Error() string {
	return e.msg
}

// Sentinel is a simple error without any annotated context, like stack trace.
// Useful for creating sentinel errors.
//
// 	const ErrSomething = errors.Sentinel("something went wrong")
//
// See https://dave.cheney.net/2016/04/07/constant-errors
type Sentinel string

func (e Sentinel) Error() string {
	return string(e)
}

// New returns a new error annotated with stack trace at the point New is called.
//
// New is a shorthand for:
// 	WithStack(NewPlain(message))
func New(message string) error {
	return WithStackDepth(NewPlain(message), 1)
}

// NewWithDetails returns a new error annotated with stack trace at the point NewWithDetails is called,
// and the supplied details.
func NewWithDetails(message string, details ...interface{}) error {
	return WithDetails(WithStackDepth(NewPlain(message), 1), details...)
}

// Errorf returns a new error with a formatted message and annotated with stack trace at the point Errorf is called.
//
// 	err := errors.Errorf("something went %s", "wrong")
func Errorf(format string, a ...interface{}) error {
	return WithStackDepth(NewPlain(fmt.Sprintf(format, a...)), 1)
}

// WithStack annotates err with a stack trace at the point WithStack was called.
// If err is nil, WithStack returns nil.
//
// WithStack is commonly used with sentinel errors and errors returned from libraries
// not annotating errors with stack trace:
//
// 	var ErrSomething = errors.NewPlain("something went wrong")
//
// 	func doSomething() error {
// 		return errors.WithStack(ErrSomething)
// 	}
func WithStack(err error) error {
	return WithStackDepth(err, 1)
}

// WithStackDepth annotates err with a stack trace at the given call depth.
// Zero identifies the caller of WithStackDepth itself.
// If err is nil, WithStackDepth returns nil.
//
// WithStackDepth is generally used in other error constructors:
//
// 	func MyWrapper(err error) error {
// 		return WithStackDepth(err, 1)
// 	}
func WithStackDepth(err error, depth int) error {
	if err == nil {
		return nil
	}

	return &withStack{
		error: err,
		stack: callers(depth + 1),
	}
}

// WithStackIf behaves the same way as WithStack except it does not annotate the error with a stack trace
// if there is already one in err's chain.
func WithStackIf(err error) error {
	return WithStackDepthIf(err, 1)
}

// WithStackDepthIf behaves the same way as WithStackDepth except it does not annotate the error with a stack trace
// if there is already one in err's chain.
func WithStackDepthIf(err error, depth int) error {
	if err == nil {
		return nil
	}

	var st interface{ StackTrace() errors.StackTrace } // nolint: unused
	if !As(err, &st) {
		return &withStack{
			error: err,
			stack: callers(depth + 1),
		}
	}

	return err
}

type withStack struct {
	error
	*stack
}

func (w *withStack) Cause() error  { return w.error }
func (w *withStack) Unwrap() error { return w.error }

// nolint: errcheck
func (w *withStack) Format(s fmt.State, verb rune) {
	switch verb {
	case 'v':
		if s.Flag('+') {
			fmt.Fprintf(s, "%+v", w.error)
			w.stack.Format(s, verb)
			return
		}
		fallthrough
	case 's':
		io.WriteString(s, w.Error())
	case 'q':
		fmt.Fprintf(s, "%q", w.Error())
	}
}

// WithMessage annotates err with a new message.
// If err is nil, WithMessage returns nil.
//
// WithMessage is useful when the error already contains a stack trace, but adding additional info to the message
// helps in debugging.
//
// Errors returned by WithMessage are formatted slightly differently:
// 	%s	error messages separated by a colon and a space (": ")
// 	%q	double-quoted error messages separated by a colon and a space (": ")
// 	%v	one error message per line
// 	%+v	one error message per line and stack trace (if any)
func WithMessage(err error, message string) error {
	return errors.WithMessage(err, message)
}

// WithMessagef annotates err with the format specifier.
// If err is nil, WithMessagef returns nil.
//
// WithMessagef is useful when the error already contains a stack trace, but adding additional info to the message
// helps in debugging.
//
// The same formatting rules apply as in case of WithMessage.
func WithMessagef(err error, format string, a ...interface{}) error {
	return errors.WithMessagef(err, format, a...)
}

// Wrap returns an error annotating err with a stack trace
// at the point Wrap is called, and the supplied message.
// If err is nil, Wrap returns nil.
//
// Wrap is a shorthand for:
// 	WithStack(WithMessage(err, message))
func Wrap(err error, message string) error {
	return WithStackDepth(WithMessage(err, message), 1)
}

// Wrapf returns an error annotating err with a stack trace
// at the point Wrapf is called, and the format specifier.
// If err is nil, Wrapf returns nil.
//
// Wrapf is a shorthand for:
// 	WithStack(WithMessagef(err, format, a...))
func Wrapf(err error, format string, a ...interface{}) error {
	return WithStackDepth(WithMessagef(err, format, a...), 1)
}

// WrapIf behaves the same way as Wrap except it does not annotate the error with a stack trace
// if there is already one in err's chain.
//
// If err is nil, WrapIf returns nil.
func WrapIf(err error, message string) error {
	return WithStackDepthIf(WithMessage(err, message), 1)
}

// WrapIff behaves the same way as Wrapf except it does not annotate the error with a stack trace
// if there is already one in err's chain.
//
// If err is nil, WrapIff returns nil.
func WrapIff(err error, format string, a ...interface{}) error {
	return WithStackDepthIf(WithMessagef(err, format, a...), 1)
}

// WrapWithDetails returns an error annotating err with a stack trace
// at the point WrapWithDetails is called, and the supplied message and details.
// If err is nil, WrapWithDetails returns nil.
//
// WrapWithDetails is a shorthand for:
// 	WithDetails(WithStack(WithMessage(err, message, details...))
func WrapWithDetails(err error, message string, details ...interface{}) error {
	return WithDetails(WithStackDepth(WithMessage(err, message), 1), details...)
}

// WrapIfWithDetails returns an error annotating err with a stack trace
// at the point WrapIfWithDetails is called, and the supplied message and details.
// If err is nil, WrapIfWithDetails returns nil.
//
// WrapIfWithDetails is a shorthand for:
// 	WithDetails(WithStackIf(WithMessage(err, message, details...))
func WrapIfWithDetails(err error, message string, details ...interface{}) error {
	return WithDetails(WithStackDepthIf(WithMessage(err, message), 1), details...)
}
